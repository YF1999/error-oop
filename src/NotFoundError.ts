import { ErrorOptions, NotFoundErrorMessageProps, NotFoundErrorProps } from './CommonTypes';
import { AbstractError } from './Native';

export abstract class AbstractNotFoundError extends AbstractError {
    protected _entityName?: string;

    public constructor(props: NotFoundErrorProps, options: ErrorOptions<NotFoundErrorMessageProps>) {
        const { entityName } = props;
        const { generateMessage: gm, ...others } = options;

        super(props, {
            generateMessage: gm && ((_props) => gm({ ..._props, entityName })),
            ...others,
        });

        this._entityName = entityName;
        this._setNonEnumerable('_entityName');
    }

    public get entityName(): string | undefined {
        return this._entityName;
    }
}

/**
 * Applicable when an attempt to retrieve data yielded no result.
 */
export class NotFoundError extends AbstractNotFoundError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param entityName The entity which is not found.
     */
    public constructor(message: string, entityName: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param entityName The entity which is not found.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, entityName: string, innerError: Error);

    public constructor(message: string = '', arg1?: string | Error, arg2?: Error) {
        // message + innerError?
        if (arg1 === undefined || typeof arg1 !== 'string') {
            super({ message, innerError: arg1 }, {});
        }

        // message + entityName + innerError?
        else {
            super({ message, entityName: arg1, innerError: arg2 }, {});
        }
    }
}
