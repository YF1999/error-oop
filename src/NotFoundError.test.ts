import { NotFoundError } from './NotFoundError';

const message = 'error';
const entityName = 'device D';

testError(
    'NotFoundError',
    () => {
        const err1 = new NotFoundError();
        const err2 = new NotFoundError(message);
        const err3 = new NotFoundError(message, err2);
        const err4 = new NotFoundError(message, entityName, err3);
        const err5 = new NotFoundError({ message, innerError: err4 });
        const err6 = new NotFoundError({ message, innerError: err5, entityName });

        return [err1, err2, err3, err4, err5, err6] as const;
    },
    (getErrors) => {
        test('property: message', () => {
            const [err1, ...errs] = getErrors();
            expect(err1.message).toBe('');
            errs.map((err) => expect(err.message).toBe(message));
        });

        test('property: inner error', () => {
            const [err1, err2, err3, err4, err5, err6] = getErrors();
            expect(err1.innerError).toBeNull();
            expect(err2.innerError).toBeNull();
            expect(err3.innerError).toBe(err2);
            expect(err4.innerError).toBe(err3);
            expect(err5.innerError).toBe(err4);
            expect(err6.innerError).toBe(err5);
        });

        test('property: entity name', () => {
            const [err1, err2, err3, err4, err5, err6] = getErrors();
            expect(err1.entityName).toBeNull();
            expect(err2.entityName).toBeNull();
            expect(err3.entityName).toBeNull();
            expect(err4.entityName).toBe(entityName);
            expect(err5.entityName).toBeNull();
            expect(err6.entityName).toBe(entityName);
        });
    },
);
