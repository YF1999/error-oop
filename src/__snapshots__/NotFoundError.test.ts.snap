// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`NotFoundError\` own keys 1`] = `
Array [
  "stack",
  "_entityName",
]
`;

exports[`Test \`NotFoundError\` own keys 2`] = `
Array [
  "stack",
  "message",
  "_entityName",
]
`;

exports[`Test \`NotFoundError\` own keys 3`] = `
Array [
  "stack",
  "message",
  "_entityName",
]
`;

exports[`Test \`NotFoundError\` own keys 4`] = `
Array [
  "stack",
  "message",
  "_entityName",
]
`;

exports[`Test \`NotFoundError\` own keys 5`] = `
Array [
  "stack",
  "message",
  "_entityName",
]
`;

exports[`Test \`NotFoundError\` own keys 6`] = `
Array [
  "stack",
  "message",
  "_entityName",
]
`;
